### Fitting Logistic Regression

In this first notebook, you will be fitting a logistic regression model to a dataset where we would like to predict if a transaction is fraud or not.

To get started let's read in the libraries and take a quick look at the dataset.

In [21]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import warnings
warnings.filterwarnings('ignore')

df = pd.read_csv('Statistics_Dataset/fraud_dataset.csv')
df.head()

transaction_id   duration      day  fraud
0           28891  21.302600  weekend  False
1           61629  22.932765  weekend  False
2           53707  32.694992  weekday  False
3           47812  32.784252  weekend  False
4           43455  17.756828  weekend  False

`1.` As you can see, there are two columns that need to be changed to dummy variables.  Replace each of the current columns to the dummy version.  Use the 1 for `weekday` and `True`, and 0 otherwise.  Use the first quiz to answer a few questions about the dataset.

In [22]:
df['weekday'] = pd.get_dummies(df['day'])['weekday']
df[['not_fraud', 'fraud']] = pd.get_dummies(df['fraud'])
df.drop(['day', 'not_fraud'], axis=1, inplace=True)

df['intercept'] = 1

df.head()

transaction_id   duration  fraud  weekday  intercept
0           28891  21.302600      0        0          1
1           61629  22.932765      0        0          1
2           53707  32.694992      0        1          1
3           47812  32.784252      0        0          1
4           43455  17.756828      0        0          1

In [23]:
print(df['fraud'].mean())
print(df['weekday'].mean())
print(df.groupby('fraud').mean()['duration'])

0.012168770612987604
0.3452746502900034
fraud
0    30.013583
1     4.624247
Name: duration, dtype: float64


`2.` Now that you have dummy variables, fit a logistic regression model to predict if a transaction is fraud using both day and duration.  Don't forget an intercept!  Use the second quiz below to assure you fit the model correctly.

In [25]:
log_model = sm.Logit(df['fraud'], df[['intercept', 'weekday', 'duration']])
results = log_model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: inf
         Iterations 16


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  fraud   No. Observations:                 8793
Model:                          Logit   Df Residuals:                     8790
Method:                           MLE   Df Model:                            2
Date:                Fri, 21 Feb 2020   Pseudo R-squ.:                     inf
Time:                        09:50:50   Log-Likelihood:                   -inf
converged:                       True   LL-Null:                        0.0000
Covariance Type:            nonrobust   LLR p-value:                     1.000
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
intercept      9.8709      1.944      5.078      0.000       6.061      13.681
weekday        2.5465      0.904      2.816      0.005       0.774       4.319
duration      -1.4637      0.290     -5.039      0.000      -2.033      -0.894
==============================================================================

Possibly complete quasi-separation: A fraction 0.98 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

In [26]:
np.exp(2.5465), np.exp(-1.4637)

(12.762357271496972, 0.2313785882117941)

In [27]:
1/np.exp(-1.4637)

4.321921089278333

### We can conclude as following:
1. Fraud is 12.76 times as likely on weekdays than weekends holding all else constant.
2. For each 1 unit increase in duration, fraud is 0.23 times as likely holding all else constant.
Alternatively, For each 1 unit decrease in duration, fruad is 4.32 times as likely holding all else constant.